### ISSUE TITLE
"java.home" does not work on macOS with one than one JDK installed

### ISSUE THREAD

**Issue by justshot** [microsoft/vscode-java-pack#339]
Issue Type: <b>Bug</b>

Setup 'java.home' does not work in neither workspace's settings.json nor user's settings.json. I have two JDK installed on my laptop. In settings.json, I point 'java.home' to my JDK 11 folder. However, VS Code always use my JDK 8 for compile/build/run/debug. Only after I deleted JDK 8, VS Code begin to use my JDK 11. 

I noticed that even I don't setup 'java.home', VS Code can still auto-detect my installed JDK and use it.

I don't have JAVA_HOME setup in my environment variables.

VS Code version: 1.43.1
macOS version: 10.13.6

Extension version: 0.8.1
VS Code version: Code 1.43.1 (fe22a9645b44368865c0ba92e2fb881ff1afce94, 2020-03-18T07:01:41.240Z)
OS version: Darwin x64 17.7.0

**Comment by Eskibear** [comment]
`java.home` overwrites others. Maybe you set the wrong value of `java.home`.  Can you check by opening `Command Palette -> Java: Configure Java Runtime`? You'll see how it determine the JDK version. 

E.g.  My default is 13, by setting java.home I can use JDK 11.
![Screen Shot 2020-03-23 at 12 16 14 PM](https://user-images.githubusercontent.com/2351748/77280789-44cdff80-6d00-11ea-9adb-696c26597973.png)

**Comment by justshot** [comment]
From ''Java: configure Java runtime", the JDK used by VS Code is Java 11
(see the first picture). However, the actual JDK used during runtime is JDK
8, which is print out by this line of code:
>
> System.out.println(MessageFormat.format("Java version: {0}", System.
> getProperty("java.version")));
>

Here is the output of the code

>  Java version: 1.8.0_162


[image: image.png]

**Comment by Eskibear** [comment]
Runtime Jdk should be specified by Java debugger extension. @testforstephen has this feature been released? Could you point out the instruction?

**Comment by justshot** [comment]
Hi

Not only runtime JDK has the problem, compile time JDK also has the same
problem. I have compile error as I am using JDK 11 classes, e.g.,
HttpClient.

**Comment by testforstephen** [comment]
@Eskibear Yes, it's released. The debugger will use the mapped JRE runtime to launch your application.

@justshot `java.home` is just used to start the Java extension itself, it doesn't means using it to run your application. If you want to use some specific JDK to compile and run your application, you need specify the Java version in your project config file such as pom.xml, build.gradle. For example, update the target java version to 11 in your project, then the extension will re-map a correct JRE for you.

**Comment by justshot** [comment]
Hi Jinbo,

You are right. The runtime JDK of my application is configured in pom.xml
in my case.

Thanks for helping!

Do you think 'java.home' and 'java:configure Java runtime" in VS code is
misleading to developers?

**Comment by testforstephen** [comment]
Agree, it should also display the actual Java runtime your project is mapped to.

//@akaroml what do you think about?

**Comment by justshot** [comment]
As a developer, what I care more is my own application's JDK setup. The JDK
used by the Java Extension Pack itself is not the key concern for
application development.

**Comment by steve78b** [comment]
I had the same problem on Windows 10, using openJDK11 

Fixed by changing pom.xml to include
        <maven.compiler.source>11</maven.compiler.source>
        <maven.compiler.target>11</maven.compiler.target>

**Comment by akaroml** [comment]
Thank you all for sharing your thoughts. We are revisiting the JDK related experience. There are two JDKs related to this problem.
1. The JDK to start the Java language server
2. The JDK that the current project is targeting
Those two can be the same or different. We'll improve the acquisition guide to provide consolidated view of both settings.

**Comment by dixitk13** [comment]
I have a similar issue or seems similar to me. Just installed the latest JDK-14. Somehow the extension is compiling with 14.x and running with 10.x. 

```
/Library/Java/JavaVirtualMachines/jdk-10.0.1.jdk/Contents/Home/bin/java -Dfile.encoding=UTF-8 @/var/folders/_9/rsq7b3sn2p104lpkq68xgp2w0000gn/T/cp_4bt0obu76nlny79zb7n5e997.argfile leet.trees.BinaryTreeLongestConsecutiveSeq 
Error: LinkageError occurred while loading main class leet.trees.BinaryTreeLongestConsecutiveSeq
        java.lang.UnsupportedClassVersionError: leet/trees/BinaryTreeLongestConsecutiveSeq has been compiled by a more recent version of the Java Runtime (class file version 58.0), this version of the Java Runtime only recognizes class file versions up to 54.0
```

My runtime/paths seems to be correct i.e. 14.x:
```
➜  git:(master) ✗ java --version
java 14.0.2 2020-07-14
Java(TM) SE Runtime Environment (build 14.0.2+12-46)
Java HotSpot(TM) 64-Bit Server VM (build 14.0.2+12-46, mixed mode, sharing)
git:(master) ✗ javac --version
javac 14.0.2
```

![image](https://user-images.githubusercontent.com/8632659/89383414-48f82980-d6b1-11ea-8a51-a9260666e7dd.png)

Also tried setting a default runtime in settings like so:
```
"java.configuration.runtimes": [
    {
      "name": "JavaSE-10",
      "path": "/Library/Java/JavaVirtualMachines/jdk-10.0.1.jdk/Contents/Home",
      "default": false
    },
    {
      "name": "JavaSE-14",
      "path": "/Library/Java/JavaVirtualMachines/jdk-14.0.2.jdk/Contents/Home",
      "default": true
    }
  ]
```

**Comment by akaroml** [comment]
> I have a similar issue or seems similar to me. Just installed the latest JDK-14. Somehow the extension is compiling with 14.x and running with 10.x.
> 
> ```
> /Library/Java/JavaVirtualMachines/jdk-10.0.1.jdk/Contents/Home/bin/java -Dfile.encoding=UTF-8 @/var/folders/_9/rsq7b3sn2p104lpkq68xgp2w0000gn/T/cp_4bt0obu76nlny79zb7n5e997.argfile leet.trees.BinaryTreeLongestConsecutiveSeq 
> Error: LinkageError occurred while loading main class leet.trees.BinaryTreeLongestConsecutiveSeq
>         java.lang.UnsupportedClassVersionError: leet/trees/BinaryTreeLongestConsecutiveSeq has been compiled by a more recent version of the Java Runtime (class file version 58.0), this version of the Java Runtime only recognizes class file versions up to 54.0
> ```
> 
> My runtime/paths seems to be correct i.e. 14.x:
> 
> ```
> ➜  git:(master) ✗ java --version
> java 14.0.2 2020-07-14
> Java(TM) SE Runtime Environment (build 14.0.2+12-46)
> Java HotSpot(TM) 64-Bit Server VM (build 14.0.2+12-46, mixed mode, sharing)
> git:(master) ✗ javac --version
> javac 14.0.2
> ```
> 
> ![image](https://user-images.githubusercontent.com/8632659/89383414-48f82980-d6b1-11ea-8a51-a9260666e7dd.png)
> 
> Also tried setting a default runtime in settings like so:
> 
> ```
> "java.configuration.runtimes": [
>     {
>       "name": "JavaSE-10",
>       "path": "/Library/Java/JavaVirtualMachines/jdk-10.0.1.jdk/Contents/Home",
>       "default": false
>     },
>     {
>       "name": "JavaSE-14",
>       "path": "/Library/Java/JavaVirtualMachines/jdk-14.0.2.jdk/Contents/Home",
>       "default": true
>     }
>   ]
> ```

@testforstephen could you please take a look?

**Comment by testforstephen** [comment]
@dixitk13 
Could you share your project's actual language level. You can open a Java file, hover at the icon below and share what it displays? 
![image](https://user-images.githubusercontent.com/14052197/90472455-2525f080-e153-11ea-984e-93f5146bbb67.png)

**Comment by dixitk13** [comment]
Looks like it still points to `jdk-10.0.1`. I tried commenting from `settings.json` > `java.configuration.runtimes` but still the same error. 
<img width="1298" alt="Screen Shot 2020-09-01 at 10 42 16 PM" src="https://user-images.githubusercontent.com/8632659/91936343-7c27db80-eca4-11ea-8c7d-62fb6767d94b.png">

**Comment by testforstephen** [comment]
@dixitk13 I guess your project doesn't use any build tool, right? You may try to clean the workspace via "**F1 -> Java: Clean the Java language server workspace**".

**Comment by dixitk13** [comment]
Nope, this particular project doesn't have any build tool ATM. Cleaning the workspace didn't help 😞

**Comment by testforstephen** [comment]
@dixitk13 What's your Java extension version? Such as Debugger and Language Support for Java.

![image](https://user-images.githubusercontent.com/14052197/91941161-55f94e80-ed2b-11ea-8a1d-51566751149d.png)

**Comment by dixitk13** [comment]
![image](https://user-images.githubusercontent.com/8632659/91941894-d9288f00-ecae-11ea-8abd-20699d843497.png)

and vscode version:
```
Version: 1.49.0-insider
Commit: b6d4fec3714ff897f72c1dab8cc02b7421a8f131
Date: 2020-09-01T05:48:29.841Z
Electron: 9.2.1
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Darwin x64 19.6.0
```

**Comment by testforstephen** [comment]
Yes, all are the latest version.

Looks like the language server used JDK 14 to compile your project, but the project runtime it resolved still points to JDK 10. This is a little weird. Would you mind sharing a sample project for us to reproduce it?

Current workaround is commenting out the below part from the `java.configuration.runtimes` to unblock you first.
```
    {
      "name": "JavaSE-10",
      "path": "/Library/Java/JavaVirtualMachines/jdk-10.0.1.jdk/Contents/Home",
      "default": false
    },
```
